### PR TITLE
Wrap up betas

### PR DESCRIPTION
**Pull request by pabloT97**
## Scope

- To implement a k-means algorithm that uses a quantum distance as a metric (beta-2)
- To wrap up, clean and add tests to beta-1 model.
- To implement minor changes in quantum distance.

## Notes

- There is  an utils.py file with some helping functions for the model. My idea is that once the models are wrapped up we can have only one utils.py file common to all the models. For the moment I am keeping one utils file per model in order to avoid merging conflicts. 
- Nothing to review in the file use_beta_1.py, as the pipelines will change.

### CODE REVIEW

File: tests/test_beta_1.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+    def test_predictions_are_integers(self)-> None:
+        """
+        Check that the predictions output by the model are integers.
+        """
+        for pred in self.beta_1.test_predictions:
+            self.assertIs(type(pred), np.int64)
+            
+
+if __name__ == '__main__':
```
Comment (ryk-wolf) [inline tests/test_beta_1.py:85]: Are we sure we want a main in a test file?

File: tests/test_beta_1.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+            'test_train_distances', './'
+        )
+        self.beta_1.load_test_train_distances(
+            'test_train_distances.pickle', './'
+        )
+        self.assertEqual(
+            distances_before_saving, self.beta_1.test_train_distances
+        )
+
+    def test_predictions_are_integers(self)-> None:
```
Comment (ryk-wolf) [inline tests/test_beta_1.py:77]: very good

File: tests/test_beta_1.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        """
+        Check that for each train instance, the length of the list 
+        with the closest vectors indexes has length equal to k. 
+        """
+        for indexes in self.beta_1.closest_vectors_indexes:
+            self.assertEqual(
+                len(indexes), args.k
+            )
+
+    def test_save_and_load_train_test_distances(self)-> None:
```
Comment (ryk-wolf) [inline tests/test_beta_1.py:61]: This is a good test but it needs a better name

File: tests/test_beta_2.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        """
+        for i in range(self.beta_2.n_its):
+            self.assertEqual(
+                len(self.beta_2.centers[i]),
+                args.k
+            )
+
+    def test_wce_is_float(self)-> None:
+        """
+        Check that wce values are a float.
```
Comment (ryk-wolf) [inline tests/test_beta_2.py:71]: Pls explicit what WCE is

File: tests/test_quantum_distance.py [hunk 5]
```diff
@@ -26,40 +41,71 @@ def test_state_preparation(self):
...
+                x1 = pad_vector_with_zeros(x1)
+                x2 = pad_vector_with_zeros(x2)
+                quantum_distance = qd(x1, x2).compute_quantum_distance()
                 euclidean_distance = np.linalg.norm(x2 - x1)
                 self.assertAlmostEqual(quantum_distance, euclidean_distance)
 
 
-
-
 if __name__ == '__main__':
```
Comment (ryk-wolf) [inline tests/test_quantum_distance.py:80]: ?

File: tests/test_quantum_distance.py [hunk 6]
```diff
@@ -26,40 +41,71 @@ def test_state_preparation(self):
...
-            X1 = qkn.normalise_vector(X1)
-            X2 = qkn.normalise_vector(X2)
-            X1 = qkn.pad_zeros_vector(X1)
-            X2 = qkn.pad_zeros_vector(X2)
-            for i in range(100):
-                x1 = X1[i]
-                x2 = X2[i]
-                model = qd(x1, x2)
-                quantum_distance = model.compute_quantum_distance()
+    def test_quantum_distance_equals_eucl_distance_padding(self)-> None:
```
Comment (ryk-wolf) [inline tests/test_quantum_distance.py:61]: Good test idea AND good name!

File: tests/test_quantum_distance.py [hunk 7]
```diff
@@ -1,21 +1,36 @@
...
 class TestQuantumDistance(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls)-> None:
+        """
+        Set up the class for testing.
+        """
+        np.random.seed(args.seed)
 
-    def test_state_preparation(self):
+    def test_proper_state_preparation(self)-> None:
```
Comment (ryk-wolf) [inline tests/test_quantum_distance.py:25]: Super vague. What does a "proper state prep" mean? Pls make that explicit in the docstring and develop the name a bit

File: neasqc_wp61/models/quantum/beta/beta_2.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+                    )
+            self.total_wce.append(wce)
+
+    def compute_predictions(self)-> None:
+        """
+        For each iteration of the algorithm training process, compute 
+        the predictions of the trainining instances, i.e.,
+        what the index of their closest cluster is. 
+        """
+        self.predictions = [
```
Comment (ryk-wolf) [inline neasqc_wp61/models/quantum/beta/beta_2.py:140]: What are these lines trying to accomplish and why?

File: tests/test_beta_1.py (hunk 1 above)
Comment (pabloT97) [reply tests/test_beta_1.py:85]: You can see how a main is used in the docs of unittest : 

https://docs.python.org/3/library/unittest.html#basic-example

File: tests/test_beta_2.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+        """
+        for i in range(self.beta_2.n_its):
+            self.assertIs(
+                type(self.beta_2.total_wce[i]),
+                float
+            )
+
+
+if __name__ == '__main__':
```
[the remark this answers is not in the document]
Comment (pabloT97) [reply tests/test_beta_2.py:80]: same reply

File: tests/test_quantum_distance.py (hunk 5 above)
Comment (pabloT97) [reply tests/test_quantum_distance.py:80]: same comment for the main as above

File: neasqc_wp61/models/quantum/beta/beta_2.py (hunk 8 above)
Comment (pabloT97) [reply neasqc_wp61/models/quantum/beta/beta_2.py:140]: The library that we are using only computes the prediction for the last iteration of the algorithm. This function computes the predictions for all the iterations during the training process

File: tests/test_beta_1.py (hunk 1 above)
Comment (ryk-wolf) [reply tests/test_beta_1.py:89]: Ok, sorry, what I said wasn't very clear. Let me rephrase my question: why do you have such an extensive main with argument parsing for so many parameters?
Are you typing them in the command line every time you run a test? Or did you add them in case someone wanted to parameterise the tests and you use the default every time? or some other way of doing things?
I looked for a note on this in the readme but didn't find one, maybe it's there and I didn't see it.

Comment (pabloT97) [reply tests/test_beta_1.py:89]: The parameters are there with the aim of not hardcoding parameters, and make things generic if possible. It was one of the comments in previous PRs. 

You are not typing them every time you run a test. You can see there how a default value is specified, which is going to be used if the user don't specify anything in the command line when running the test file. It is exactly the same way as we use to do it in the pipelines (use_{model}.py) for the report.

Comment (ryk-wolf) [reply tests/test_beta_1.py:89]: I see where this all came from now!
I've set up a quick meeting to discuss this on Monday :).

File: neasqc_wp61/models/quantum/beta/beta_2.py (hunk 8 above)
Comment (ryk-wolf) [reply neasqc_wp61/models/quantum/beta/beta_2.py:140]: You're creating a list of lists of None.
I don't understand why you'd want to do that.
Could you explain why you need a list of lists of None?

I you really want to create a list of lists of None, there are more optimised ways to do so (you can lookup things related to repeat).

Otherwise you can just create it in list comprehension by changing a few of the things you have below, no need to create a big None object before populating it, you can create it directly!

**Review by ryk-wolf — CHANGES_REQUESTED** [review]
I have seen several of my comments marked as addressed and not being addressed. They do need to be addressed.
Pls make sure you don't inadvertently mark something as addressed before you've actually made the changes

File: neasqc_wp61/models/quantum/beta/beta_2.py (hunk 8 above)
Comment (pabloT97) [reply neasqc_wp61/models/quantum/beta/beta_2.py:140]: The reason why I need a list of None is because I need to I will add elements to it in a "disordered" way. In the loop in lines 149-152, I will add elements so that for example in the first iteration of the loop, I add what would be the fourth element of the list. After looking up here : 

https://stackoverflow.com/questions/50835478/a-way-to-insert-a-value-in-an-empty-list-in-a-specific-index

it seems that the only way of doing it is with a list of None or with a dictionary. I decided to use a list for the sake of consistency with the rest of the models. 

I changed the way I create it, so that it is slightly more optimised. I looked up the repeat function of itertools library (I am not sure if u were referring to it in your suggestion) but the code looks very similar to the solution I came up with, so I decided not to use itertools in order to avoid importing the library.

File: tests/test_beta_2.py [hunk 10]
```diff
@@ -1,69 +1,133 @@
...
-                random_initialisation = True,
-                seed = args.seed
-            )
-        else : 
-            cls.beta_2.initialise_cluster_centers(
-                random_initialisation=False)
-        cls.beta_2.train_k_means_algorithm()
-        cls.beta_2.compute_wce()
-        cls.beta_2.compute_predictions()
+        pass
```
Comment (ryk-wolf) [inline tests/test_beta_2.py:101]: ???

File: tests/test_quantum_distance.py [hunk 11]
```diff
@@ -109,56 +169,23 @@ def test_quantum_distance_equals_eucl_distance(self)-> None:
...
+                if i == 1:
+                    vector_1 = pad_vector_with_zeros(vector_1)
+                    vector_2 = pad_vector_with_zeros(vector_2)
+                quantum_distance = qd(
+                    vector_1,vector_2
+                ).compute_quantum_distance()
+                euclidean_distance = np.linalg.norm(vector_2 - vector_1)
+                self.assertAlmostEqual(
+                    quantum_distance,
+                    euclidean_distance
```
Comment (ryk-wolf) [inline tests/test_quantum_distance.py:186]: This is a very long test. I feel you're doing 2 tests in 1, am i correct? The first with fixed vectors, the second not. I'd split this into 2 tests. You can say you test that specific vector pair in the first one and the other is a more general case.

File: tests/test_quantum_distance.py [hunk 12]
```diff
@@ -69,30 +109,50 @@ def test_value_error_is_raised_if_vectors_dont_have_norm_1(
...
+        self
+    )-> None:
+        """
+        For two vectors with same length, test that a ValueError is
+        raised when computing the quantum distance if one or both of
+        them have norm different from 1.
+        """
+        for j in range(self.n_samples -1):
+            vector_1 = self.vectors_that_need_padding[j]
+            vector_1 = normalise_vector(vector_1)
```
Comment (ryk-wolf) [inline tests/test_quantum_distance.py:130]: You generally do not want to be using your own library to make operations when you're testing for smth.
If this test fails, it could be either because my `normalise_vector` function is broken or because my `compute_quantum_distance` one is not throwing the error => I don't know why the test fails => this test is not useful.

If you need to normalise stuff, use a sturdy lib like numpy to do that.

In general test results shoul not depend on other functions than the one they're testing.

File: tests/test_quantum_distance.py (hunk 11 above)
Comment (ryk-wolf) [reply tests/test_quantum_distance.py:186]: Also, as noted in another test, try to remove as many internal functions from this code as you can. Limit inter-dependencies!

File: tests/test_quantum_distance.py [hunk 13]
```diff
@@ -13,36 +13,77 @@
...
-                        -args.x_limit, args.x_limit, size = size_vector_1
-                    )
-                )
-                cls.vectors_that_need_padding[i].append(
-                    np.random.uniform(
-                        -args.x_limit, args.x_limit, size = size_vector_2
-                    )
-                )
-        np.random.seed(args.seed)
+        pass
```
Comment (ryk-wolf) [inline tests/test_quantum_distance.py:86]: ???

File: tests/test_beta_1.py [hunk 14]
```diff
@@ -1,45 +1,111 @@
...
+    'seed' : 30031935,
+    'n_vectors_train' : 25,
+    'n_vectors_test' : 10,
+    'vectors_limit_value' : 1000,
+    'n_classes_limit' : 5,
+    'vectors_size_limit' : 8,
+    'k_limit' : 5
+}
+
+def set_up_class_attributes(test_args : dict)-> list:
```
[the remark this answers is not in the document]
Comment (pabloT97) [reply tests/test_beta_1.py:25]: I am sorry but it's not very clear to me which names are you referring to . Which x, y ? I have changed them for vectors and labels in all files.

File: neasqc_wp61/models/quantum/beta/beta_1.py [hunk 15]
```diff
@@ -18,50 +18,52 @@ class QuantumKNearestNeighbours:
...
 
     def compute_test_train_distances(self)-> None:
         """
         Compute and store the quantum distance
         between each pair of test/train instances.
         """
         self.test_train_distances = [[] for _ in range(self.n_test)]
-        for i,xte in enumerate(self.x_test):
-            for xtr in self.x_train:
+        for i,xte in enumerate(self.vectors_test):
```
[the remark this answers is not in the document]
Comment (pabloT97) [reply neasqc_wp61/models/quantum/beta/beta_1.py:58]: My bad! forgot to change that... addresed now

File: tests/test_quantum_distance.py (hunk 11 above)
Comment (pabloT97) [reply tests/test_quantum_distance.py:186]: Split the test in two. 

For the internal functions see comment below.

File: tests/test_quantum_distance.py (hunk 12 above)
Comment (pabloT97) [reply tests/test_quantum_distance.py:130]: Completely agree with what you say.  

I was looking up very quickly some libraries but I could'nt  find any that does exactly what we want. However you can see the functions, they are super simple: 

-normalise_vector is just one line doing : vector/np.linalg.norm(vector), so it's actually using numpy in a very simple way.
-pad_vector_with_zeros pads with zeros until the vector length is a power of two. No function in numpy does directly that.

If you want I can spend more time looking for libraries that has functions doing exactly what we want. but that of course will take more time...

File: tests/test_beta_1.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (pabloT97) [reply tests/test_beta_1.py:25]: right! my bad!

**Review by ryk-wolf — APPROVED** [review]
